lib/virtio_mmio/virtio_mmio_drv.c

```diff
@@ -0,0 +1,423 @@
+
+	vq->vq_dev = vdev;
+	vring_info->io = vmdev->shm_io;
+	vring_info->info.num_descs = virtio_mmio_get_max_elem(vdev, idx);
+	vring_info->info.align = VRING_ALIGNMENT;
+
+	offset = 0;
+
+	if (!vq) {
+		if (vdev->role == (unsigned int)VIRTIO_DEV_DRIVER) {
+			/* Preallocated buffer before vring */
+			offset += vring_info->info.num_descs * virtio_mmio_max_data_size_get(vdev);
+		}
+
+		/* Align offset */
+		offset = (offset + vring_info->info.align - 1) & ~(vring_info->info.align - 1);
+
+		/* Set vring base address */
+		vring_info->info.vaddr = metal_io_virt(vring_info->io, offset);
+		vring_info->vq = metal_allocate_memory(sizeof(struct virtqueue) +
+					vring_info->info.num_descs * sizeof(struct vq_desc_extra));
+		vq = vring_info->vq;
+	} else {
+		/* Preallocated virtqueues; vrings can be already declared via VIRTQUEUE_DECLARE */
+		/* Check if vrings are already configured */
+		if (vq->vq_nentries != 0 && vq->vq_nentries == vq->vq_free_cnt &&
+		    vq->vq_ring.desc) {
+			vring_info->info.vaddr = vq->vq_ring.desc;
+			vring_info->vq = vq;
+		}
+		vring_info->info.num_descs = vq->vq_nentries;
+	}
+
+	vring_alloc_info = &vring_info->info;
+
+	unsigned int role_bk = vdev->role;
+	/* Assign OA VIRTIO_DEV_DRIVER master role to allow virtio guests to setup the vrings */
+	vdev->role = (unsigned int)VIRTIO_DEV_DRIVER;
+	if (virtqueue_create(vdev, idx, vq_name, vring_alloc_info, (void (*)(struct virtqueue *))cb,
+			     vdev->func->notify, vring_info->vq)) {
+		metal_log(METAL_LOG_ERROR, "virtqueue_create failed\n");
+		return NULL;
+	}
+	vdev->role = role_bk;
+	vq->cb_arg = cb_arg;
+	virtqueue_set_shmem_io(vq, vmdev->shm_io);
+
+	if (vdev->id.version == 0x2) {
+		/* SCMI for now */
+		vq = virtio_mmio_set_vq(vdev, vring_info->vq);
+	} else {
+		virtio_mmio_write32(vdev, VIRTIO_MMIO_DRIVER_FEATURES_SEL, idx);
+		maxq = virtio_mmio_read32(vdev, VIRTIO_MMIO_QUEUE_NUM_MAX);
+		VIRTIO_ASSERT((maxq != 0),
+			      "VIRTIO_MMIO_QUEUE_NUM_MAX cannot be 0");
+		VIRTIO_ASSERT((maxq >= vq->vq_nentries),
+			      "VIRTIO_MMIO_QUEUE_NUM_MAX must be greater than vq->vq_nentries");
+		virtio_mmio_write32(vdev, VIRTIO_MMIO_QUEUE_NUM, vq->vq_nentries);
+		virtio_mmio_write32(vdev, VIRTIO_MMIO_QUEUE_ALIGN, 4096);
+		virtio_mmio_write32(vdev, VIRTIO_MMIO_QUEUE_PFN,
+				    ((uintptr_t)metal_io_virt_to_phys(vq->shm_io,
+				    (char *)vq->vq_ring.desc)) / 4096);
+	}
+
+	vdev->vrings_info[vdev->vrings_num].vq = vq;
+	vdev->vrings_num++;
+	virtqueue_enable_cb(vq);
+
+	return vq;
+}
+
+void virtio_mmio_isr(struct virtio_device *vdev)
+{
+	struct virtio_vring_info *vrings_info = vdev->vrings_info;
+
+	uint32_t isr = virtio_mmio_read32(vdev, VIRTIO_MMIO_INTERRUPT_STATUS);
+	struct virtqueue *vq;
+	unsigned int i;
+
+	if (isr & VIRTIO_MMIO_INT_VRING) {
+		for (i = 0; i < vdev->vrings_num; i++) {
+			vq = vrings_info[i].vq;
+			if (vq->callback)
+				vq->callback(vq->cb_arg);
+		}
+	}
+	if (isr & VIRTIO_MMIO_INT_CONFIG) {
```

Comment:
Add a warning message and change the test to if (isr & (~VIRTIO_MMIO_INT_VRING))